src/react/components.js

```diff
@@ -294,16 +302,45 @@ export function applyGetDerivedStateFromProps(
           invariant(b instanceof Value);
           return AbstractValue.createFromConditionalOp(realm, condition, a, b);
         }
+      } else if (kind === "||" || kind === "&&") {
+        let a = deriveState(state.args[0]);
+        let b = deriveState(state.args[1]);
+        if (b === null) {
+          invariant(a instanceof Value);
+          return AbstractValue.createFromLogicalOp(realm, kind, a, prevState);
+        } else {
+          invariant(a instanceof Value);
+          invariant(b instanceof Value);
+          return AbstractValue.createFromLogicalOp(realm, kind, a, b);
+        }
       } else {
-        invariant(false, "TODO: unknown abstract value passed to deriveState");
+        invariant(state.args !== undefined, "TODO: unknown abstract value passed to deriveState");
+        // as the value is completely abstract, we need to add a bunch of
+        // operations to be emitted to ensure we do the right thing at runtime
+        let a = AbstractValue.createFromBinaryOp(realm, "!==", state, realm.intrinsics.null);
+        let b = AbstractValue.createFromBinaryOp(realm, "!==", state, realm.intrinsics.undefined);
+        let c = AbstractValue.createFromLogicalOp(realm, "&&", a, b);
+        invariant(c instanceof AbstractValue);
+        let newState = new ObjectValue(realm, realm.intrinsics.ObjectPrototype);
+        // we cannot use the standard Object.assign as partial state
+        // is not simple. however, given getDerivedStateFromProps is
+        // meant to be pure, we can assume that there are no getters on
+        // the partial abstract state
```

Comment:
I'd like to have @sebmarkbage weigh on this. Still not sure this is necessarily true although it makes sense.

Reply:
Hm. This is an interesting case.

Objects are free to have getters even in pure functions as long as they only mutate things created by that function.

```js
({
  _foo: null;
  get lazy() {
     if (!this._foo) { this._foo = compute(); }
     return this._foo;
  }
})
```

That's why we're resilient to these in abstract function calls which might invoke them.

What makes this case particularly interesting is that the getter is not invoked until after it leaves the pure function. So it is React itself that is causing this effect. Not the user code.

If this mutates something that was returned from state, then that already existed at the time of the call. So the read call is not pure, which makes that intermediate operation not pure.

There exists a pure operation in between getDerivedStateFromProps (pure), and render (pure). I guess we can call it the "getMergedState" function. And if there is a getter like this on the partial state object, then that makes that hidden function not pure. That breaks the rules of React.

So I think this is technically in line with our rules to assume it doesn't have mutating getters.

Now whether it can throw is another issue.

Reply:
It is fine if the getter is invoked inside gDSFP because that's the function that created it.

The issue is that it leaks and gets called by the functional getMergedState function. The getter wasn't created by the getMergedState function so in that context it breaks the rules.

Another way to break the rules is accessing the getter in render.

```js
getDerivedStateFromProps() {
  return {
    _foo: null;
    get lazy() {
       if (!this._foo) { this._foo = compute(); }
       return this._foo;
    }
  };
}

render() {
  var x = this.state.lazy; // side-effect in render
}
```